test/extensions/filters/http/jwt_authn/extractor_test.cc

```diff
@@ -0,0 +1,186 @@
+  }
+
+  JwtAuthentication config_;
+  ExtractorConstPtr extractor_;
+};
+
+TEST_F(ExtractorTest, TestNoToken) {
+  auto headers = TestHeaderMapImpl{};
+  auto tokens = extractor_->extract(headers);
+  EXPECT_EQ(tokens.size(), 0);
+}
+
+TEST_F(ExtractorTest, TestWrongHeaderToken) {
+  auto headers = TestHeaderMapImpl{{"wrong-token-header", "jwt_token"}};
+  auto tokens = extractor_->extract(headers);
+  EXPECT_EQ(tokens.size(), 0);
+}
+
+TEST_F(ExtractorTest, TestWrongParamToken) {
+  auto headers = TestHeaderMapImpl{{":path", "/path?wrong_token=jwt_token"}};
+  auto tokens = extractor_->extract(headers);
+  EXPECT_EQ(tokens.size(), 0);
+}
+
+TEST_F(ExtractorTest, TestDefaultHeaderLocation) {
+  auto headers = TestHeaderMapImpl{{"Authorization", "Bearer jwt_token"}};
+  auto tokens = extractor_->extract(headers);
+  EXPECT_EQ(tokens.size(), 1);
+
+  EXPECT_EQ(tokens[0]->token(), "jwt_token");
+  EXPECT_TRUE(tokens[0]->isIssuerSpecified("issuer1"));
+
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer2"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer3"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer4"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer5"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("unknown_issuer"));
+
+  // Test token remove
+  tokens[0]->removeJwt(headers);
+  EXPECT_FALSE(headers.Authorization());
+}
+
+TEST_F(ExtractorTest, TestDefaultParamLocation) {
+  auto headers = TestHeaderMapImpl{{":path", "/path?access_token=jwt_token"}};
+  auto tokens = extractor_->extract(headers);
+  EXPECT_EQ(tokens.size(), 1);
+
+  EXPECT_EQ(tokens[0]->token(), "jwt_token");
+  EXPECT_TRUE(tokens[0]->isIssuerSpecified("issuer1"));
+
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer2"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer3"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer4"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer5"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("unknown_issuer"));
+}
+
+TEST_F(ExtractorTest, TestCustomHeaderToken) {
+  auto headers = TestHeaderMapImpl{{"token-header", "jwt_token"}};
+  auto tokens = extractor_->extract(headers);
+  EXPECT_EQ(tokens.size(), 1);
+
+  EXPECT_EQ(tokens[0]->token(), "jwt_token");
+  EXPECT_TRUE(tokens[0]->isIssuerSpecified("issuer2"));
+  EXPECT_TRUE(tokens[0]->isIssuerSpecified("issuer4"));
+
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer1"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer3"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("issuer5"));
+  EXPECT_FALSE(tokens[0]->isIssuerSpecified("unknown_issuer"));
+
+  // Test token remove
+  tokens[0]->removeJwt(headers);
+  EXPECT_FALSE(headers.get(Http::LowerCaseString("token-header")));
+}
+
+TEST_F(ExtractorTest, TestPrefixHeaderNotMatch) {
+  auto headers = TestHeaderMapImpl{{"prefix-header", "jwt_token"}};
+  auto tokens = extractor_->extract(headers);
+  EXPECT_EQ(tokens.size(), 0);
+}
+
+TEST_F(ExtractorTest, TestPrefixHeaderMatch) {
```

Comment:
Would prefer comments above each `TEST_F` explaining what they do for the uninitiated.

Reply:
Done